accepted/exception-handling.md

```diff
@@ -0,0 +1,540 @@
+the on-disk encoding into some in-memory data structure, which would
+slow down `Module` creation. We can make it contain only and exactly
+the information we actually need. We can use a format that is fast and
+easy to decode. All this results in smaller tables, faster lookups,
+and slimmer `.cwasm` binaries. While we won't be able to piggyback on
+tools like `dwarfdump` &mdash; given how little information we need to
+store in the side tables and the relative simplicity that brings, and
+the fact that we can always use `std::fmt::Debug` in tactically-placed
+logs &mdash; the lack of `dwarfdump` isn't the end of the
+world.[^dwarf-future]
+
+[^fp-unwind-future]: Whether we stop using frame pointers and start
+    recovering the previous frame via a side table is an orthogonal
+    question and something we can change later. On the one hand, it
+    would be nice to recover the FP and get another register for
+    x64. On the other hand, frame pointers make the unwinder a lot
+    simpler, which is generally pretty tricky code already, and is
+    also critical for correctness in the face of GC references. Frame
+    pointers also make integration with system tooling (profilers,
+    debuggers, etc...) dead simple and generally Just Work. But either
+    way, we don't need or plan on changing anything about this at the
+    moment, and can revisit this decision at some time in the future;
+    don't have to worry about it now.
+
+[^dwarf-future]: All that said, if we ever decide to switch from
+    frame-pointer stack walking to side-table-based stack walking, it
+    may make sense to switch to a DWARF subset at that point in time,
+    since we would start using all that extra information that FDEs
+    can and must encode, unlike today. Doing that switch at that
+    hypothetical point in the future is fine because everything here
+    is an internal implementation detail and definitely not something
+    that is guaranteed across Wasmtime versions.
+
+### Incremental Milestone: Exceptions as Uncatchable Traps
+
+We are interested in landing this feature in a reasonably timely
+manner to enable toolchain producers to use exception handling and
+allow dependent proposals (such as stack switching) to run early
+experiments on top of Wasmtime.
+
+Therefore we propose to start by implementing exceptions as
+uncatchable traps. This will force implementation of the initial
+`wasmtime::Config` knobs, CLI flags, some Wasm-to-CLIF translation,
+etc... It allows toolchains, like Kotlin, to start running subsets of
+their test suite on Wasmtime, even if they use exceptions-related
+instructions, as long as the dynamic execution doesn't actually throw
+any exceptions.
+
+This is also a mode that we may want to support indefinitely anyways:
+some programs "must" be compiled with exceptions due to accidental
+complexity with (say) their build system, but won't ever actually
+raise any exceptions due to the configuration of various knobs. With
```

Comment:
I feel like wasm-opt or such should get a mode which turns all exception throws into traps. That way other wasm engines without exception support would also be able to handle these modules.

Reply:
I agree that would be nice, but of course I don't maintain `wasm-opt` :)

We can always remove this knob if people don't use it or don't find it useful. But it will be nice as an incremental milestone either way. And it would always be useful if `wasm-opt` didn't support some proposed Wasm feature that Wasmtime did, and therefore you couldn't use `wasm-opt` to do this as a wasm-to-wasm transform because it would balk at the unimplemented feature usage.